### PR TITLE
feat: source build mode

### PR DESCRIPTION
**Pull request by targeral**
### <samp>🤖 Generated by Copilot at 9dc6956</samp>

This pull request introduces a new `plugin-source-build` package that enables source code build mode for monorepo projects. The package provides a builder plugin that modifies the webpack config to resolve modules based on the closest `tsconfig.json` file. The pull request also adds support for the plugin in the `app-tools` package and updates the `package.json` script and the `turbo` cache. The pull request also deletes an unused file and adds some ESLint and vitest configurations for the new package.

### <samp>🤖 Generated by Copilot at 9dc6956</samp>

*  Add a new builder plugin `plugin-source-build` to enable source code build mode for monorepo projects (F5-F16)

## Checklist

- [ ] I have added changeset via `pnpm run change`.
- [ ] I have updated the documentation.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/toolkit/monorepo-utils/package.json [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  },
+  "scripts": {
+    "prepublishOnly": "only-allow-pnpm",
+    "new": "modern-lib new",
+    "dev": "modern-lib build --watch",
+    "build": "modern-lib build",
+    "test": "jest --passWithNoTests"
+  },
+  "dependencies": {
+    "@microsoft/rush-lib": "^5.99.0",
```
Comment (chenjiahan) [inline packages/toolkit/monorepo-utils/package.json:43]: `@microsoft/rush-lib` contains so many sub-dependencies, maybe we should move it to devDependencies.

File: packages/toolkit/monorepo-utils/package.json [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    "@modern-js/utils": "^2",
```
Comment (chenjiahan) [inline packages/toolkit/monorepo-utils/package.json:44]:
```suggestion
    "@modern-js/utils": "workspace:*",
```

File: packages/builder/builder-webpack-provider/src/config/validate/experiments.ts [hunk 3]
```diff
@@ -2,4 +2,8 @@ import { sharedExperimentsConfigSchema, z } from '@modern-js/builder-shared';
 import type { ExperimentsConfig } from '../../types';
 
 export const experimentsConfigSchema: z.ZodType<ExperimentsConfig> =
-  sharedExperimentsConfigSchema;
+  sharedExperimentsConfigSchema
+    .extend({
+      sourceBuild: z.boolean(),
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/config/validate/experiments.ts:7]: If the `sourceBuild` config is built in, we should also make `builder-plugin-source-build` as a built-in plugin.

But `builder-plugin-source-build` now has some third-party dependencies, we do not want to increase the install size of builder, so we need to reduce the dependencies if we want to build in it.

Comment (targeral) [reply packages/builder/builder-webpack-provider/src/config/validate/experiments.ts:7]: I think some third-party dependencies should be required, such as @pnpm/filter-workspace-packages, @pnpm/logger. So we turn `builder-plugin-source-build` into an external plugin?

File: packages/toolkit/monorepo-utils/package.json (hunk 1 above)
Comment (targeral) [reply packages/toolkit/monorepo-utils/package.json:43]: With these considerations in mind, perhaps we should use builder-plugin-build-source as an external plugin.

File: packages/toolkit/monorepo-utils/package.json (hunk 2 above)
Comment (targeral) [reply packages/toolkit/monorepo-utils/package.json:44]: OK

File: packages/builder/builder-webpack-provider/src/config/validate/experiments.ts (hunk 3 above)
Comment (targeral) [reply packages/builder/builder-webpack-provider/src/config/validate/experiments.ts:7]: I have removed some third-party dependencies from builder-plugin-source-build

File: packages/toolkit/monorepo-utils/package.json (hunk 1 above)
Comment (targeral) [reply packages/toolkit/monorepo-utils/package.json]: `@microsoft/rush-lib` has been removed from the project.

File: packages/builder/builder-webpack-provider/src/config/validate/experiments.ts (hunk 3 above)
Comment (chenjiahan) [reply packages/builder/builder-webpack-provider/src/config/validate/experiments.ts:7]: Yes, we can turn `builder-plugin-source-build` into an external plugin now.

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**